### PR TITLE
refactor: use native Java threads in `parMap`

### PR DESCRIPTION
**Pull request by sockmaster27**

### CODE REVIEW

File: main/test/ca/uwaterloo/flix/util/TestParOps.scala [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+package ca.uwaterloo.flix.util
+
+import ca.uwaterloo.flix.api.Flix
+import org.scalatest.funsuite.AnyFunSuite
+
+import java.util.concurrent.Executors
+
+class TestParOps extends AnyFunSuite {
+
+  implicit var flix: Flix = new Flix()
```
Comment (magnus-madsen) [inline main/test/ca/uwaterloo/flix/util/TestParOps.scala:25]: val?

**Comment by magnus-madsen** [comment]
Can you try `Xperf` and `--Xbenchmark-throughput` to check that performance is about the same?

**Comment by sockmaster27** [comment]
====== master ======
```
Throughput (best): 30.001 lines/sec (with 16 threads.)

  min:  4.749, max: 30.001, avg: 22.573, median: 25.631

  The highest throughput was in iteration: 6 (out of 10).
  The ratio between the best and worst iteration was: 6,3x.

Finished 10 iterations on 66.136 lines of code in 39 seconds.
```


====== new ======
```
Throughput (best): 27.977 lines/sec (with 16 threads.)

  min:  4.921, max: 27.977, avg: 22.856, median: 27.517

  The highest throughput was in iteration: 8 (out of 10).
  The ratio between the best and worst iteration was: 5,7x.

Finished 10 iterations on 66.136 lines of code in 38 seconds.
```

**Comment by sockmaster27** [comment]
====== master ======
```
Throughput (best): 34.832 lines/sec (with 16 threads.)

  min: 28.304, max: 34.832, avg: 31.480, median: 32.166

Finished 7 iterations on 66.136 lines of code in 14 seconds.
```


====== new ======
```
Throughput (best): 32.967 lines/sec (with 16 threads.)

  min: 30.725, max: 32.967, avg: 31.807, median: 32.055

Finished 7 iterations on 66.136 lines of code in 14 seconds.
```

**Comment by magnus-madsen** [comment]
!!!jar

**Comment by magnus-madsen** [comment]
I also get lower performance. But no idea why?

**Comment by magnus-madsen** [comment]
Can we try to use a https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ForkJoinPool.html and then use the `execute` method on it? This should safe some allocations of futures.

**Comment by magnus-madsen** [comment]
I would use this constructor: https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ForkJoinPool.html#ForkJoinPool-int- inside Flix

**Comment by sockmaster27** [comment]
I've now tried using `.execute()` both on the threadpool and the forkjoinpool, but no matter what, running Xperf seems to just land on a best somewhere random between 31.000 and 37.000 lines per second.

**Comment by sockmaster27** [comment]
So for now I'm still just using the threadpool.

**Comment by magnus-madsen** [comment]
So-- for unknown reasons -- Scala ParArray remains faster??

**Comment by sockmaster27** [comment]
> So-- for unknown reasons -- Scala ParArray remains faster??

It seems to be varying wildly, but maybe(???) ParArray is a little more consistent towards the middle. It's very noisy and I think at this point the differences from run to run are completely overshadowing the differences between the implementations.

**Comment by magnus-madsen** [comment]
> > So-- for unknown reasons -- Scala ParArray remains faster??
> 
> It seems to be varying wildly, but maybe(???) ParArray is a little more consistent towards the middle. It's very noisy and I think at this point the differences from run to run are completely overshadowing the differences between the implementations.

You can bump the N in CompilerPerf. I can also try when I get home.

I wonder if the countdown latch is the bottleneck.

**Comment by magnus-madsen** [comment]
But is the code using ForkJoinPool??

**Comment by sockmaster27** [comment]
> But is the code using ForkJoinPool??

I tried both, they were identical, so I stayed with the simpler thread pool.

**Comment by magnus-madsen** [comment]
!!!jar

**Comment by magnus-madsen** [comment]
See https://github.com/flix/flix/pull/6690

**Pull request closed without merging**